 <h1 class="firstHeading" id="firstHeading">Ngoại lệ</h1> <p>Xử lý ngoại lệ cung cấp một cách chuyển kiểm soát và thông tin từ một số điểm trong việc thực hiện một chương trình sang một trình xử lý được liên kết với một điểm được thông qua trước đó bởi việc thực hiện (nói cách khác, ngoại lệ xử lý việc chuyển giao kiểm soát ngăn xếp cuộc gọi).</p>
<p>Một ngoại lệ có thể được ném bởi một<a href="throw" title="cpp/language/throw">throw-expression</a>Thì<a href="dynamic_cast" title="cpp/language/dynamic cast"><code>dynamic_cast</code></a>Thì<a href="typeid" title="cpp/language/typeid"><code>typeid</code></a>Thì<a href="new" title="cpp/language/new">new-expression</a>Thì<a href="../memory/new/operator_new" title="cpp/memory/new/operator new">allocation function</a>và bất kỳ chức năng thư viện tiêu chuẩn nào được chỉ định để ném các ngoại lệ để báo hiệu các điều kiện lỗi nhất định (ví dụ:<code><a href="../container/vector/at" title="cpp/container/vector/at">std::vector::at</a></code>Thì<code><a href="../string/basic_string/substr" title="cpp/string/basic string/substr">std::string::substr</a></code>, vân vân).</p>
<p>Để một ngoại lệ bị bắt, biểu hiện ném phải ở bên trong một<a href="try_catch" title="cpp/language/try catch">try-block</a>hoặc bên trong một chức năng được gọi từ một khối thử, và phải có một<a href="try_catch" title="cpp/language/try catch">catch clause</a>phù hợp với loại đối tượng ngoại lệ.</p>
<p>Khi khai báo một hàm, (các) thông số kỹ thuật sau đây có thể được cung cấp để giới hạn các loại ngoại lệ mà một hàm có thể ném:</p>
<table class="t-rev-begin"> <tr class="t-rev t-until-cxx17">
<td> <ul><li> <a href="except_spec" title="cpp/language/except spec">dynamic exception specifications</a> </li></ul> </td> <td><span class="t-mark-rev t-until-cxx17">xáo trộn</span></td>
</tr> </table> <table class="t-rev-begin"> <tr class="t-rev t-since-cxx11">
<td> <ul><li> <a href="noexcept_spec" title="cpp/language/noexcept spec">noexcept specifications</a> </li></ul> </td> <td><span class="t-mark-rev t-since-cxx11">(kể từ C ++ 11)</span></td>
</tr> </table> <p>Lỗi phát sinh trong quá trình xử lý ngoại lệ được xử lý bởi<code><a href="../error/terminate" title="cpp/error/terminate">std::terminate</a></code><span class="t-rev-inl t-until-cxx17"><span>Và<code><a href="../error/unexpected" title="cpp/error/unexpected">std::unexpected</a></code></span><span><span class="t-mark-rev t-until-cxx17">xáo trộn</span></span></span>.</p>
<h3 id="Usage">Cách sử dụng</h3> <p>Mặc dù biểu hiện ném có thể được sử dụng để chuyển kiểm soát sang một khối mã tùy ý lên ngăn xếp thực thi, vì lý do tùy ý (tương tự như<code><a href="../utility/program/longjmp" title="cpp/utility/program/longjmp">std::longjmp</a></code>), việc sử dụng dự định của nó là xử lý lỗi.</p>
<h4 id="Error_handling">Xử lý lỗi</h4> <p>Ném một ngoại lệ được sử dụng để báo hiệu các lỗi từ các chức năng, trong đó "lỗi" thường chỉ giới hạn ở<sup class="reference" id="cite_ref-1"><a href="#cite_note-1">[1]</a></sup><sup class="reference" id="cite_ref-2"><a href="#cite_note-2">[2]</a></sup><sup class="reference" id="cite_ref-3"><a href="#cite_note-3">[3]</a></sup>:</p>
<ol>
<li>Thất bại trong việc đáp ứng các điều kiện hậu, chẳng hạn như không tạo ra một đối tượng giá trị trả về hợp lệ.</li>
<li>Thất bại trong việc đáp ứng các điều kiện tiên quyết của một chức năng khác phải được gọi.</li>
<li>.</li>
</ol> <p>Cụ thể, điều này ngụ ý rằng sự thất bại của các hàm tạo (xem thêm<a href="raii" title="cpp/language/raii">RAII</a>) và hầu hết các nhà khai thác nên được báo cáo bằng cách ném ngoại lệ.</p>
<p>Ngoài ra, cái gọi là<i>hợp đồng rộng</i>Các chức năng sử dụng các ngoại lệ để chỉ ra các đầu vào không thể chấp nhận được, ví dụ,<code><a href="../string/basic_string/at" title="cpp/string/basic string/at">std::string::at</a></code>Không có điều kiện tiên quyết, nhưng ném một ngoại lệ để chỉ ra chỉ mục ngoài phạm vi.</p>
<h4 id="Exception_safety">An toàn ngoại lệ</h4> <p>Sau khi điều kiện lỗi được báo cáo bởi một hàm, các đảm bảo bổ sung có thể được cung cấp liên quan đến trạng thái của chương trình. Bốn cấp độ đảm bảo ngoại lệ sau đây thường được công nhận<sup class="reference" id="cite_ref-4"><a href="#cite_note-4">[4]</a></sup><sup class="reference" id="cite_ref-5"><a href="#cite_note-5">[5]</a></sup><sup class="reference" id="cite_ref-6"><a href="#cite_note-6">[6]</a></sup>, đó là những supersets nghiêm ngặt của nhau:</p>
<ol>
<li> <i>Nothrow (hoặc NoFail) Đảm bảo ngoại lệ</i>- Hàm không bao giờ ném ngoại lệ. Nothrow (lỗi được báo cáo bằng các phương tiện khác hoặc che giấu) được mong đợi<a href="destructor" title="cpp/language/destructor">destructors</a>và các chức năng khác có thể được gọi trong quá trình tháo gỡ.<span class="t-rev-inl t-since-cxx11"><span>Các<a href="destructor" title="cpp/language/destructor">destructors</a>&gt;<a href="noexcept" title="cpp/language/noexcept"><code>noexcept</code></a>Theo mặc định.</span><span><span class="t-mark-rev t-since-cxx11">(kể từ C ++ 11)</span></span></span>Nofail (chức năng luôn thành công) được dự kiến ​​là hoán đổi,<a href="move_constructor" title="cpp/language/move constructor">move constructors</a>và các chức năng khác được sử dụng bởi các chức năng cung cấp bảo đảm ngoại lệ mạnh mẽ.</li>
<li> <i>Đảm bảo ngoại lệ mạnh mẽ</i>- Nếu hàm ném một ngoại lệ, trạng thái của chương trình sẽ được chuyển trở lại trạng thái ngay trước khi gọi chức năng (ví dụ,<code><a href="../container/vector/push_back" title="cpp/container/vector/push back">std::vector::push_back</a></code>).</li>
<li> <i>Đảm bảo ngoại lệ cơ bản</i>- Nếu chức năng ném một ngoại lệ, chương trình ở trạng thái hợp lệ. Không có tài nguyên nào bị rò rỉ, và tất cả các bất biến của các đối tượng đều còn nguyên vẹn.</li>
<li> <i>Không có bảo đảm ngoại lệ</i>-Nếu hàm ném một ngoại lệ, chương trình có thể không ở trạng thái hợp lệ: rò rỉ tài nguyên, tham nhũng bộ nhớ hoặc các lỗi phá hủy bất biến khác có thể xảy ra.</li>
</ol> <p>Ngoài ra, các thành phần chung có thể<i>Đảm bảo trung lập ngoại lệ</i>: Nếu một ngoại lệ được ném từ tham số mẫu (ví dụ: từ<code>Compare</code>đối tượng chức năng của<code><a href="../algorithm/sort" title="cpp/algorithm/sort">std::sort</a></code>hoặc từ người xây dựng<code>T</code>TRONG<code><a href="../memory/shared_ptr/make_shared" title="cpp/memory/shared ptr/make shared">std::make_shared</a></code>), nó được tuyên truyền, không thay đổi, cho người gọi.</p>
<h3 id="Exception_objects">Đối tượng ngoại lệ</h3> <p>Mặc dù các đối tượng thuộc bất kỳ loại hình hoàn chỉnh nào và con trỏ CV vào void có thể được ném dưới dạng các đối tượng ngoại lệ, tất cả các hàm thư viện tiêu chuẩn đều ném các đối tượng tạm thời ẩn danh theo giá trị và các loại của các đối tượng đó đều có nguồn gốc (trực tiếp hoặc gián tiếp) từ<code><a href="../error/exception" title="cpp/error/exception">std::exception</a></code>. Các ngoại lệ do người dùng xác định thường theo mẫu này.<sup class="reference" id="cite_ref-7"><a href="#cite_note-7">[7]</a></sup><sup class="reference" id="cite_ref-8"><a href="#cite_note-8">[8]</a></sup><sup class="reference" id="cite_ref-9"><a href="#cite_note-9">[9]</a></sup></p>
<p>Để tránh sao chép không cần thiết của đối tượng ngoại lệ và cắt đối tượng, cách thực hành tốt nhất cho các điều khoản bắt là bắt được bằng cách tham khảo.<sup class="reference" id="cite_ref-10"><a href="#cite_note-10">[10]</a></sup><sup class="reference" id="cite_ref-11"><a href="#cite_note-11">[11]</a></sup><sup class="reference" id="cite_ref-12"><a href="#cite_note-12">[12]</a></sup><sup class="reference" id="cite_ref-13"><a href="#cite_note-13">[13]</a></sup></p>
<h3 id="External_links">Liên kết bên ngoài</h3> <table> <tr style="vertical-align:top;"> <td> </td> <td>
<ol class="references"> <li id="cite_note-1"> <span class="reference-text">H. Sutter (2004)<a class="external text" href="https://www.drdobbs.com/when-and-how-to-use-exceptions/184401836" rel="nofollow">"When and How to Use Exceptions"</a>Trong Tiến sĩ Dobb's</span> </li> <li id="cite_note-2"> <span class="reference-text">H. Sutter, A. Alexandrescu (2004), "Tiêu chuẩn mã hóa C ++", Mục 70</span> </li> <li id="cite_note-3"> <span class="reference-text">Hướng dẫn cốt lõi C ++<a class="external text" href="https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ri-except" rel="nofollow">I.10: Use exceptions to signal a failure to perform a required task</a></span> </li> <li id="cite_note-4"> <span class="reference-text">B. Stroustrup (2000), "Ngôn ngữ lập trình C ++"<a class="external text" href="https://stroustrup.com/3rd_safe.pdf" rel="nofollow">Appendix E</a></span> </li> <li id="cite_note-5"> <span class="reference-text">H. Sutter (2000) "C ++ đặc biệt"</span> </li> <li id="cite_note-6"> <span class="reference-text">D. Áp -ra -ham (2001)<a class="external text" href="https://www.boost.org/community/exception_safety.html" rel="nofollow">"Exception Safety in Generic Components"</a></span> </li> <li id="cite_note-7"> <span class="reference-text">D. Áp -ra -ham (2001)<a class="external text" href="https://www.boost.org/community/error_handling.html" rel="nofollow">"Error and Exception Handling"</a></span> </li> <li id="cite_note-8"> <span class="reference-text">isocpp.org Super-Faq<a class="external text" href="https://isocpp.org/wiki/faq/exceptions#what-to-throw" rel="nofollow">"What should I throw?"</a></span> </li> <li id="cite_note-9"> <span class="reference-text">Hướng dẫn cốt lõi C ++<a class="external text" href="https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-exception-types" rel="nofollow">E.14: Use purpose-designed user-defined types as exceptions (not built-in types)</a></span> </li> <li id="cite_note-10"> <span class="reference-text">Hướng dẫn cốt lõi C ++<a class="external text" href="https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-exception-ref" rel="nofollow">E.15: Throw by value, catch exceptions from a hierarchy by reference</a></span> </li> <li id="cite_note-11"> <span class="reference-text">S. Meyers (1996) "C ++ hiệu quả hơn" Mục 13</span> </li> <li id="cite_note-12"> <span class="reference-text">isocpp.org Super-Faq<a class="external text" href="https://isocpp.org/wiki/faq/exceptions#what-to-catch" rel="nofollow">"What should I catch?"</a></span> </li> <li id="cite_note-13"> <span class="reference-text">H. Sutter, A. Alexandrescu (2004) "Tiêu chuẩn mã hóa C ++" Mục 73</span> </li> </ol> </td>
</tr>
</table> <div class="_attribution">
<p class="_attribution-p">© CPPReference.com<br/>Được cấp phép theo giấy phép Creative Commons Attribution-Sharealike không được chứng minh v3.0.<br/>
<a class="_attribution-link" href="https://en.cppreference.com/w/cpp/language/exceptions">https://en.cppreference.com/w/cpp/language/exceptions</a>
</p>
</div>
